pkg/driver_cryptocell_310/psa_cryptocell_310/cipher_chacha20.c

```diff
@@ -0,0 +1,259 @@
+}
+
+psa_status_t psa_cipher_chacha20_set_iv(psa_cipher_chacha20_ctx_t *ctx,
+                                        const uint8_t *iv,
+                                        size_t iv_length)
+{
+    DEBUG("Peripheral ChaCha20 Cipher set IV");
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    CRYSError_t periph_status = CRYS_CHACHA_GENERAL_ERROR;
+    switch (iv_length)
+    {
+        case 8:
+            status = PSA_ERROR_NOT_SUPPORTED;
+            break;
+        case 12:
+            cryptocell_310_enable();
+            periph_status = CRYS_CHACHA_Init(   &ctx->ctx.ctx,
+                                                (uint8_t*)iv,
+                                                CRYS_CHACHA_Nonce96BitSize,
+                                                ctx->ctx.pre_setup.key,
+                                                0,
+                                                ctx->ctx.pre_setup.mode);
+            cryptocell_310_disable();
+            status = CRYS_to_psa_error(periph_status);
+            break;
+        case 16:
+            periph_status = CRYS_CHACHA_Init(   &ctx->ctx.ctx,
+                                                &((uint8_t*)iv)[4],
+                                                CRYS_CHACHA_Nonce96BitSize,
+                                                ctx->ctx.pre_setup.key,
+                                                unaligned_get_u32(iv),
+                                                ctx->ctx.pre_setup.mode);
+            cryptocell_310_disable();
+            status = CRYS_to_psa_error(periph_status);
+            break;
+        default:
+            status = PSA_ERROR_INVALID_ARGUMENT;
+    }
+    return status;
+}
+
+psa_status_t psa_cipher_chacha20_update(psa_cipher_chacha20_ctx_t *ctx,
+                                        const uint8_t *input,
+                                        size_t input_length,
+                                        uint8_t *output,
+                                        size_t output_size,
+                                        size_t *output_length)
+{
+    DEBUG("Peripheral ChaCha20 Cipher update");
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    if (output_size < input_length)
+    {
+        return PSA_ERROR_BUFFER_TOO_SMALL;
+    }
+
+    if (!ctx->is_finished)
+    {
+        const uint32_t blocks_length = (input_length / CRYS_CHACHA_BLOCK_SIZE_IN_BYTES) * CRYS_CHACHA_BLOCK_SIZE_IN_BYTES;
+        /* xcrypt full blocks */
+        cryptocell_310_enable();
+        CRYSError_t periph_status = CRYS_CHACHA_Block(  &ctx->ctx.ctx,
+                                                        (uint8_t*)input,
+                                                        blocks_length,
+                                                        output);
+        
+        if (periph_status != CRYS_OK) {
+            return CRYS_to_psa_error(periph_status);
+        }
```

Comment:
```suggestion
        status = CRYS_to_psa_error(periph_status);
        if (status != PSA_SUCCESS) {
            return status;
        }
```

I'd rather do it that way around, but maybe that's subjective.